zkevm-circuits/src/evm_circuit/execution/log.rs

```diff
@@ -0,0 +1,630 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{LOG_STATIC_GAS, N_BYTES_MEMORY_WORD_SIZE},
+        step::ExecutionState,
+        table::{CallContextFieldTag, RwTableTag, TxLogFieldTag},
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition,
+                Transition::{Delta, To},
+            },
+            from_bytes,
+            math_gadget::IsZeroGadget,
+            memory_gadget::{MemoryAddressGadget, MemoryExpansionGadget},
+            sum, Cell, Word,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+use array_init::array_init;
+use eth_types::Field;
+use eth_types::{evm_types::OpcodeId, ToLittleEndian, ToScalar};
+use halo2_proofs::{circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct LogGadget<F> {
+    same_context: SameContextGadget<F>,
+    // memory address
+    memory_address: MemoryAddressGadget<F>,
+    topics: [Cell<F>; 4],
+    is_topic_zeros: [IsZeroGadget<F>; 4],
+
+    contract_address: Cell<F>,
+    is_static_call: Cell<F>,
+    is_persistent: Cell<F>,
+    tx_id: Cell<F>,
+    memory_expansion: MemoryExpansionGadget<F, 1, N_BYTES_MEMORY_WORD_SIZE>,
+}
+
+impl<F: Field> ExecutionGadget<F> for LogGadget<F> {
+    const NAME: &'static str = "LOG";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::LOG;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        // check opcode is in [log0,log4], already do SameContextGadget::construct
+        let memory_offset = cb.query_cell();
+        let msize = cb.query_rlc();
+
+        // Pop mstart_address, msize from stack
+        cb.stack_pop(memory_offset.clone().expr());
+        cb.stack_pop(msize.expr());
+
+        // read tx id
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+
+        // constrain not in static call
+        let is_static_call = cb.call_context(None, CallContextFieldTag::IsStatic);
+        cb.require_zero("is_static_call is false", is_static_call.expr());
+
+        // check contract_address in CallContext & TxLog
+        // use call context's  callee address as contract address
+        let contract_address = cb.call_context(None, CallContextFieldTag::CalleeAddress);
+        let is_persistent = cb.call_context(None, CallContextFieldTag::IsPersistent);
+        cb.condition(is_persistent.expr(), |cb| {
+            cb.tx_log_lookup(
+                1.expr(),
+                tx_id.expr(),
+                TxLogFieldTag::Address,
+                0.expr(),
+                contract_address.expr(),
+            );
+        });
+
+        // constrain topics in logs
+        let topics = array_init(|_| cb.query_cell());
+        let is_topic_zeros: [IsZeroGadget<F>; 4] =
```

Comment:
if use bool cells, we need to do additional check whether these bools are correct, current approach take topics from stack directly and check if it is zero, and conditionally lookup RW to ensure sanity checking.

Reply:
Ok, let me write down the constraints to see if I follow:
1. topic_count = opcode - OpcodeId::LOG0
2. sum(selector_exprs) == topic_count
3. selector_exprs can only go from 0 to 0, 0 to 1 and 1 to 1. [0, 0, ..., 1, 1]
Isn't `1-selector_exprs` the expression that should be used here https://github.com/appliedzkp/zkevm-circuits/blob/dd03ade797180e178b5cb576b810c93ddd3a9b4d/zkevm-circuits/src/evm_circuit/execution/logs.rs#L81-L89
Instead of `is_topic_zeros`?
Otherwise you could be doing `LOG1(0x0)` and the stack pop and log topic lookup would be disabled, right?
I just don't understand why there's a special behavior when the topic value is 0.

Reply:
@ed255  per discussion,  not all topics are hashed 32 bytes , so change to use 4 selectors, which is bool , from 1 zero, when it is 1, lookup stack to match it.  updated in [1ab5927](https://github.com/appliedzkp/zkevm-circuits/pull/335/commits/1ab5927d527a746846f5c050b3613ea63fa65183) thanks for catching!